add --tolerate_transitive mode

fiesh opened the pull request:
This adds `--tolerate_transitive` mode as discussed in #526.

It seems that due to other issues, we still can't really use IWYU though :( (#406, #424 at least).

I think the mode is quite reasonable though and should most often improve users' experience.

kimgr commented:
Thanks! There seems to be some build error -- see Travis logs. I'll try and look into this tonight.

fiesh commented:
Ah sorry, in all rebasing madness I made some stupid mistake.  I fixed and... rebased ;)

kimgr commented on iwyu_globals.cc:
```diff
@@ -179,8 +183,10 @@ int CommandlineFlags::ParseArgv(int argc, char** argv) {
+    {"tolerate_transitive", optional_argument, nullptr, 'r'},
     {nullptr, 0, nullptr, 0}
   };
+  // TODO: This doesn't cover all options!
```
I think that's intentional, not all options are exposed as short-opts.

fiesh replied:
I'll remove it then

kimgr commented on iwyu_globals.cc:
```diff
@@ -93,6 +93,9 @@ static void PrintHelp(const char* extra_msg) {
          "   --no_fwd_decls: do not use forward declarations.\n"
+         "   --tolerate_transitive: If a file uses symbols from both a.h and b.h\n"
+         "        and includes a.h, tolerate b.h both when it is included and \n"
+	 "        when\n it is not.\n"
```
Misaligned, needs another leading space. And spurious \n after "when"

fiesh replied:
Oops, nvim put a tab in there, and I didn't see that because it aligns with tab width 8.

kimgr commented on iwyu_output.cc:
```diff
@@ -468,8 +468,8 @@ bool OneIncludeOrForwardDeclareLine::HasSymbolUse(const string& symbol_name)
 
-void OneIncludeOrForwardDeclareLine::AddSymbolUse(const string& symbol_name) {
-  ++symbol_counts_[symbol_name];
+void OneIncludeOrForwardDeclareLine::AddSymbolUse(const string& symbol_name, int multiplicity) {
```
Can you remove this method to `AddSymbolsUses` and call `multiplicity` `count` instead?

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,59 @@ void CalculateIwyuForFullUse(OneUse* use,
+// but not present but transitively included by some
+// OneIncludedOrForwardDeclareLine b that is desired and present
+// and adds a's symbol counts to b.
+vector<OneIncludeOrForwardDeclareLine> TolerateTransitively(vector<OneIncludeOrForwardDeclareLine> const & oiofdls,
```
Can you call variables of type `OneIncludeOrForwardDeclareLine` `line[s]` instead? I like the `oiofdls` acronym, but it's hard to read and spell correctly :)

kimgr replied:
`TolerateTransitively` does not describe what this method does, more a general attitude :)

I wonder what a good name would be... `RemoveTransitivelyProvided` or something? Shorter, preferably.

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,59 @@ void CalculateIwyuForFullUse(OneUse* use,
   }
 }
 
+// This function removes all OneIncludeOrForwardDeclareLine's a that are desired
```
Please reflow comment to fill 80 chars

kimgr replied:
Typo: a that

No need for the apostrophe on OneIncludeOrForwardDeclareLines

kimgr commented on iwyu_output.h:
```diff
@@ -161,7 +161,7 @@ class OneIncludeOrForwardDeclareLine {
   void clear_line_numbers() { start_linenum_ = end_linenum_ = -1; }
   // Another symbol we're using that's defined in this file.
-  void AddSymbolUse(const string& symbol_name);
+  void AddSymbolUse(const string& symbol_name, int multiplicity = 1);
```
I think I'd prefer to remove the default arg and update callers to pass 1 where applicable.

fiesh commented:
I incorporated your suggestions hopefully to your liking.  The only thing missing is the clang-format, which I'd gladly do because I'm actually not used to formatting code by hand any more.  However, I don't see and clang-format style file in the project.  Is there something?  The defaults don't seem to match.

kimgr commented:
There's a .clang-format file in the project root. Maybe it's hidden because of the leading dot?

WSoptics commented:
Ah ok heh, I had actually used that and got confused by `clang-format` still changing tons of lines.  So what should I do, only reformat my lines?  Or alter the whole file?

It would seem that everything should be uniformly formatted?  Then again, that shouldn't be in my commit I guess.

fiesh commented:
Ah ok heh, I had actually used that and got confused by clang-format still changing tons of lines. So what should I do, only reformat my lines? Or alter the whole file?

It would seem that everything should be uniformly formatted? Then again, that shouldn't be in my commit I guess.

kimgr commented:
Yes, it would be nice if you could constrain it to your changed lines. We'll have to take a closer look at global reformatting or not later.

fiesh commented:
OK, please look at the updated pull request.  Would rebase once everything is fine!

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+    CHECK_(iter != first.end() &&
+           "Internal error, cannot find the source for a transitive tolerate "
+           "target");
+    for (const auto& p : target.symbol_counts()) {
```
Better name than `p` -- `symbol_count`, maybe?

fiesh replied:
I went for `p` since it's a pair because we're iterating over a map.  Other names always have the problem that they somewhat obfuscate we're dealing with a key-value-pair.  But I can change it if you think it makes it more readable.

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+vector<OneIncludeOrForwardDeclareLine> PruneTransitivelyPresent(
+    vector<OneIncludeOrForwardDeclareLine> const& lines,
+    const IwyuPreprocessorInfo* preprocessor_info) {
+  // In the first step, we copy all lines that
```
Everything you describe here is pretty obvious from the code itself. Could you say something more about the intent and motivation of the algorithm instead? There's a lot of different states influencing the outcome here, and I'm having trouble keeping it all in my head without a higher purpose.

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+    // Check if target is transitively included by some source.
+    // This must be proper, equality does not suffice.
+    for (const OneIncludeOrForwardDeclareLine& source : lines) {
+      if (!source.IsIncludeLine()) {
```
I think you can combine these two if-statements into a single `||`-separated one.

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+    }
+    return true;
+  };
+  vector<OneIncludeOrForwardDeclareLine> first, second;
```
Blank line after the lambda, please

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+  // * or that are an include file, not present, desired, and whose include_file
+  //   isn't transitively included by another line which is an include file
+  // into one range and the remaining ones into another range.
+  const auto pred = [&lines, preprocessor_info](
```
Can you name `pred` after its purpose. It looks something like `is_transitively_present`, maybe...?

fiesh replied:
Good suggestion!

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+  second.reserve(lines.size());
+  std::partition_copy(lines.cbegin(), lines.cend(), std::back_inserter(first),
+                      std::back_inserter(second), pred);
+  // Now add the symbols from range second to the corresponding elements of
```
This could also do with some more detail. E.g. why are we doing this?

fiesh commented:
Thank you for your review!  I'll try to incorporate your suggestions and commit a new release.

I don't know why, but I cannot comment on your last comment, maybe a github bug.  So for the apostrophe between `OneIncludeOrForwardDeclareLine` and `s`, I like putting it there because it makes clear what the symbol name is and that's it's merely a plural form. I can change it if you like that, but I think it makes it worse.  I can also use markdown-like syntax and put the `OneIncludeOrForwardDeclareLine` into \`\`s.

kimgr commented:
Huh, I don't see a Reply field for the last comment either. Weird.

I've only seen apostrophes in English to indicate either the possessive form ("my cat's fur is silky") or to indicate word contraction ("don't ask"). We typically don't use backticks in comments in IWYU, single quotes are more common, but that won't look very nice here. I think you should just drop the apostrophe.

fiesh commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+    }
+    return true;
+  };
+  vector<OneIncludeOrForwardDeclareLine> first, second;
```
Not quite sure about this.  `first` and `second` have the advantage of making clear to the user what arguments they are in the call to `std::partition_copy`.  What the two ranges contain is rather difficult and explained in length further up.

fiesh commented:
I included almost all of your suggestions, I think they have improved the code a lot.  If you can have a look at it again, I'd be most thankful!

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,70 @@ void CalculateIwyuForFullUse(OneUse* use,
+    return true;
+  };
+
+  vector<OneIncludeOrForwardDeclareLine> first, second;
```
"direct"/"indirect"?

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,70 @@ void CalculateIwyuForFullUse(OneUse* use,
+  // This is because elements of range second are then precisely those whose
+  // symbols are already exposed through a different header that will is to be
+  // included anyway, so they can be omitted.
+  const auto is_transitively_present = [&lines, preprocessor_info](
```
I wonder if this name is right. I realize I suggested it, but I meant `is_transitively_present` as an example to make you describe the purpose. It looks to me like it now returns `false` for lines that have transitive representation but `true` for everything else, so is it backwards?

I'm also having trouble with the `source` and `target` terminology. I think `target` could be renamed `line` and `source` could be renamed `other`. That way it's clearer that you cross-reference every include-line against every other.

fiesh replied:
You're right, I flipped its logic, so that it does match the name `is_trasitively_present` now and accordingly switched `first` and `second`.  (And then renamed `first` and `second`.)

fiesh commented:
@kimgr Oh yes, I am.  Sorry for the delay.  It'll be a couple more days before I get to it though, I'm occupied with real life's work.

kimgr commented:
@fiesh Oh, great! I was considering picking up where you left off, but then I can focus on other things, thanks!

fiesh commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,70 @@ void CalculateIwyuForFullUse(OneUse* use,
   }
 }
 
+// This function removes all OneIncludeOrForwardDeclareLines a that are desired
```
It's not a typo, the variable name is `a`.  I'll rename it to make the explanation clearer.

fiesh commented:
I have incorporated nearly all your suggestions in the new commit.  The formatting was also fixed by applying `clang-format`, don't know why I hadn't done that before, sorry.

The only thing I did not change is the renaming of `source` and `target`.  I do feel they describe the situation in the dependency graph quite well, the `source`s being the sources of paths with the `target`s being their targets.

I would very much appreciate some first test from someone else though since I think me writing the test might lead to a false feeling of it working correctly.  I developed it using some include files whose behavior I wanted a certain way, and I think I might integrate certain assumptions.  Plus I alas am rather constrained time wise recently.  I do want to see this feature integrated though :)

Thanks for all your feedback, @kimgr

kimgr commented on iwyu_output.cc:
```diff
@@ -2007,6 +2072,10 @@ size_t IwyuFileInfo::CalculateAndReportIwyuViolations() {
   }
 
+  if (GlobalFlags().tolerate_transitive) {
+	  lines_ = internal::PruneTransitivelyPresent(lines_, preprocessor_info_);
```
Too much indentation here

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+  const auto is_transitively_present =
+      [&lines,
+       preprocessor_info](const OneIncludeOrForwardDeclareLine& target) {
+        if (!target.IsIncludeLine() || target.is_present() ||
```
`is_present` and `is_desired` are both cheaper than `IsIncludeLine`, so please put them first.

kimgr commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+  // listed as including the corresponding symbols because they do transitively.
+  for (const OneIncludeOrForwardDeclareLine& target : indirect) {
+    const auto source =
+        std::find_if(direct.begin(), direct.end(),
```
Hmm. It looks a little off that you need to search for this again -- you've already worked every include line against every other once.

It looks like you already have an O(n^2) algorithm: `partition_copy` traverses all lines, and the lambda you're using to partition traverses all lines. And now this traverses all lines invoking `FileTransitivelyIncludes` again. I wonder if you could flatten this so it only needs to do the N*N loop once?

jru replied:
Also think so. If the lambda returned a pointer to the file that transitively includes the target or null, instead of true or false, then a single pass through the lines should be enough. No extra vectors or partition/find_if would be needed either.

The lambda could then be called `transitively_included_by` or smth along those lines.

fiesh replied:
Yes good catch, that's really unnecessary!

fiesh replied:
OK, now that I've looked at it a bit, I'm not sure what the best venue would be.  Returning a pointer isn't really that good of an option because it would still require two iterations over the range: one to add the `target`s' symbols to the `source`s, and one with a `std::remove_if` or so to eliminate these `target`s then.  I don't see how to do this in one go.

One option would be to copy everything into a `vector` of `pair`s of `OneIncludeOrForwardDeclarLine`s and pointers to such, then fill the pointers, then add the symbols, then call `remove_if` on all whose pointer is a `nullptr`, and then `transform` the range back into one where the pointers are removed.  Not sure if I like that option better, also not sure if it's actually faster or slower since it involves more allocations and copying, which might be worse than doing some work twice.

Thoughts?

kimgr replied:
I need to sit down and think about this for a while. It must be possible to make this better than O(N*N) + O(N) (where `FileTransitivelyIncludes` is also non-trivial)

jru replied:
I thought briefly about this and I believe it would actually need two passes through the lines to prevent aliasing issues in its current form: one to save the <source,target>pairs and one more to process the vector, and add the symbols from each target to the corresponding source. For the first pass, only the sources that found a valid target would need to be stored (skipping the ones that return null)

That said, I can't actually try anything right now, so please take everything with a grain of salt :)

fiesh commented on iwyu_output.cc:
```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
+      source->AddSymbolUses(p.first, p.second);
+    }
+    // source is desired now because target was.
+    CHECK_(target.is_desired()) << "Target must be desired";
```
I'm not entirely certain what else to say?  `source` is the direct provider, `target` was indirect, and `target` was desired, so that now source is.

fiesh replied:
Or are you talking about the `CHECK_`?  It's just an assertion for what already happened in `is_transitively_present`.

kimgr replied:
Yes, the check message. If it ever fires, it's nice to know up-front *why* target must be desired (presumably because you only selected desired targets before?)

fiesh commented:
The new commit addresses all comments except the two that I in turn commented.  Thanks for the feedback!

kimgr commented:
I haven't had much time to experiment with this, and my personal schedule is pretty packed in the coming month, so it might be a while until I can get to it.

One other idle thought that popped up: you're discarding a lot of includes in the new `Prune` method. Should you instead just mark them as non-desired? That way we could add a later pass and still include them in the analysis (if that becoms necessary for some reason).

fiesh commented:
Oh wow!  I will look into this in about two weeks, alas I cannot do it before then.  I'm very interested!

jru commented:
I think it looks great! I added a couple comments.

kimgr commented:
Here's a stable link with @jru's `is_present` comment addressed: https://github.com/kimgr/include-what-you-use/commit/tolerate_transitive.

I would really need to write up some test cases to see if this works according to expectations...

kimgr commented:
I wrote a simple test case and debugged this. I think the big takeaway from that was to only fiddle with desred/present in the second pass, after the include map is built. The transitive graph is the same whether lines are present and desired or not.

I think the test boiled down to three distinct cases:
1) direct.h and indirect.h are both included, only indirect.h symbols are used: remove direct.h because IWYU
2) direct.h and indirect.h are both included, symbols from both are used: leave indirect.h because IWYU
3) Only direct.h is included, symbols from both are used: do nothing because tolerate-transitive

Does that jive with everybody else's assumptions? Can you see a case I've missed?

I'll turn these into real test cases and push a branch soon.

fiesh commented:
@kimgr Sorry for my late reply, I had a lot of things going on in real life.

Your three cases look correct and cover all relevant cases.

I looked at your code and find it very clean and much better readable, great job!  There's one minor recommendation from my side: Ditch the `multimap` in favor of a plain `vector`, which is then `sort`ed and `erase`d + `unique`d.  This tends to be much faster.

kimgr commented:
Posted #594 to replace this. @fiesh -- please review that version and check if it meets your needs. Closing.
